### DISCUSSION TITLE
reproducibility of builds

### DISCUSSION THREAD

**Discussion by MrBitBucket** [pypa/cibuildwheel#878] [General]
I spent an age messing with cibuildwheel for reportlab and eventually managed to get most of the wheels I wanted. Rather than mess with the main repo again I created a copy and started to run actions there. The macos builds immediately started to fail.

I looked at the logs for the macos build and see immediately some differences

working https://github.com/MrBitBucket/reportlab-mirror/runs/3882039705?check_suite_focus=true
Current runner version: '2.283.1'
Environment: macos-10.15 Version: 20211002.1
Virtual Environment Provisioner 1.0.0.0-master-20210922-1
Successfully setup CPython (3.9.7)

failing https://github.com/MrBitBucket/ci-reportlab/runs/3904225960?check_suite_focus=true
Current runner version: '2.283.2'
Environment: macos-10.15 Version: 20211011.2
Virtual Environment Provisioner 1.0.0.0-master-20211007-1
Successfully setup CPython (3.10.0)

The failing case errors out building Python 3.6.8; I assume that's because the environment is now different.

Is there any way to make these builds more specific eg what can I use instead of macos-latest etc etc?

I know little about macos so am a bit lost here. It seems to me that wheel builds should be reproducible, but github and others are constantly updating. My own experiments with a local manylinux build on different hosts show that docker does not isolate from the host system.

**Comment by MrBitBucket** [comment]
FWIW I made yet another copy and this time the action runs fine. I find no differences between the repo codes and the same macos builder and OS versions are used. The first difference I see  is that the failing case last week installs  bracex-2.1.1-py3-none-any.whl and the good test today installs bracex-2.2-py3-none-any.whl; the 2.1.1 version was fine previously so I don't think that's an issue. Mysteries of github I suppose.

**Comment by MrBitBucket** [comment]
I am a dumbo sorry for noise; seems I added a standard .gitignore creating the first copy and that carefully hid my src/reportlab/lib folder :(

Even so I would be glad to know how to fix the various versions. I see ubuntu-18.04, but can I add a .point release eg 20.04.3?

**Reply by Czaki** [reply]
Linux build is done on docker images. The host version should not impact on output wheel.

**Comment by MrBitBucket** [comment]
Thats what I thought, but comparing the actual outputs does show a difference my test script looked like this

```
export CIBW.......
cibuildwheel --output-dir="$wheelhouse"
mkdir -p "$wheels"
mv "$wheelhouse"/* "$wheels"/
```
running on arch linux with 20.10.9 and ubuntu 18.04 with docker 20.10.2 I saw differences in the test output produced by a graphics process (a warning "colinear!"). I suspect that randomness was not being added properly in the ubuntu runs. I believe the produced wheels were OK.
